lib/seattleflu/id3c/cli/command/etl/redcap_det_fh_airs.py

```diff
@@ -0,0 +1,910 @@
+        return None
+
+    return build_residential_location_resources(
+        db = db,
+        cache = cache,
+        housing_type = housing,
+        primary_street_address = enrollment['enr_mail_street_address'],
+        secondary_street_address = enrollment['enr_mail_street_address_2'],
+        city = enrollment['enr_mail_city'],
+        state = enrollment['enr_mail_state'],
+        zipcode = enrollment['enr_mail_zip'],
+        system_identifier = INTERNAL_SYSTEM)
+
+
+def airs_build_specimens(db,
+    patient_reference: dict,
+    initial_encounter_reference: dict,
+    redcap_record_instance: REDCapRecord,
+    collection_method: CollectionMethod) -> tuple:
+
+    specimen_entry = None
+    specimen_observation_entry = None
+    specimen_identifier = None
+    specimen_received = (collection_method == CollectionMethod.SWAB_AND_SEND and \
+        is_complete('post_collection_data_entry_qc', redcap_record_instance))
+
+    if specimen_received:
+        # Use barcode fields in this order.
+        # TODO: verify with study
+        prioritized_barcodes = [
+            redcap_record_instance["results_barcode"],
+            redcap_record_instance["return_utm_barcode"],
+            redcap_record_instance["pre_scan_barcode"],
+        ]
+
+        # Check if specimen barcode is valid
+        specimen_barcode = None
+        for barcode in prioritized_barcodes:
+            specimen_barcode = barcode.strip()
+            if specimen_barcode:
+                # Disable logging when calling find_identifier here to suppess alerts
+                logging.disable(logging.WARNING)
+                specimen_identifier = find_identifier(db, specimen_barcode)
+                logging.disable(logging.NOTSET)
+
+                break
+
+        if specimen_identifier:
+            specimen_entry, specimen_reference = create_specimen(
+                prioritized_barcodes = prioritized_barcodes,
+                patient_reference = patient_reference,
+                collection_date = get_collection_date(redcap_record_instance, collection_method),
+                sample_received_time = redcap_record_instance['samp_process_date'],
+                able_to_test = redcap_record_instance['able_to_test'],
+                system_identifier = INTERNAL_SYSTEM)
+
+            specimen_observation_entry = create_specimen_observation_entry(
+                specimen_reference = specimen_reference,
+                patient_reference = patient_reference,
+                encounter_reference = initial_encounter_reference)
+        else:
+            LOG.debug(f"No identifier found for barcode «{specimen_barcode}»")
+            LOG.info("No identifier found for barcode. Creating encounter for record instance without sample")
+    else:
+        LOG.debug("Creating encounter for record instance without sample")
+
+    # Do not log warning for unrecognized barcodes, which are being tracked outside of ID3C.
```

Comment:
This might be worth some discussion of how QC for unrecognized barcodes should be handled for this study.

Reply:
Yeah, this is just another holdover from the UW reopening ETL.